### PR TITLE
BigQuery import of logit CSV

### PR DESCRIPTION
**Pull request by matth79**
BigQuery is a scalable way to extract from the complete logits data, selecting by time, species, etc.

We might like to play around with BigQuery for a fun map visualization of species counts or to do ad-hoc queries to answer questions during our workshop, which might in some cases prove heavy in R.

This change has a Beam pipeline that converts CSV rows to Avro records, which is a format that makes ingest to BigQuery easier. One CSV row still corresponds to one Avro record, but the Avro record has the point number, latitude, and longitude added, and it structures the logits as a map from species code to logit instead of as a vector of logits. (Denormalization is common for BigQuery.)

Optionally, logits below a threshold or for certain species codes can be omitted.

### CODE REVIEW

File: acoustic/flatten_bird_csv/src/flatten_bird_csv.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  Returns:
+    None
+  """
+  birds_txt = os.path.join(base_dir, 'input', 'birds.txt')
+  csv_pattern = os.path.join(base_dir, 'input', '**', '*.csv')
+  output = os.path.join(base_dir, 'output', 'avro')
+
+  with beam.Pipeline(options=options) as pipeline:
+    bird_list = beam.pvalue.AsSingleton(
+        pipeline | ('ReadBirds' >> beam.io.textio.ReadFromText(birds_txt))
```
Comment (sdenton4) [inline acoustic/flatten_bird_csv/src/flatten_bird_csv.py:142]: Do these (presumably parallel-ish) operations guarantee that the output order is maintained between steps?

I would usually just read the bird list before starting the pipeline to ensure that everything is in the correct order. (But I guess that's why I'm not a Haskell programmer.)

Comment (matth79) [reply acoustic/flatten_bird_csv/src/flatten_bird_csv.py:142]: They do not guarantee order. I knew this but apparently didn't know it. Hmm...

The reason for reading the bird list in the pipeline is not so much a preference to be purely functional as it is that Python open() does not work to read files from GCS. What I should have been doing is using Beam Match transforms to sent Beam ReadableFiles, or their contents, as side inputs.

Comment (sdenton4) [reply acoustic/flatten_bird_csv/src/flatten_bird_csv.py:142]: Maybe a good option then would be to use a transform which reads the whole file and then a ParDo to parse the full contents in one go, instead of line-by-line?

Comment (ddkapan) [reply acoustic/flatten_bird_csv/src/flatten_bird_csv.py:142]: Hi M & T, 

Looking forward to pulling in this feature, sounds really flexible and powerful.

BTW: See also Mark & my R code from the first 'live' feature for this week:

~/acoustic/data_ingest

https://github.com/calacademy-research/COMB/commit/651a2744bce3e9d351180dba03e04633c1a1b5a5

it gets and puts data from here:

https://drive.google.com/drive/folders/1-BDu2u1T_S_EWdrb25iiroAL_kpWYEmS

including some precomputed logit filters that allow us moRtals to continue to schlep along coding ... ;)

Send your pull request when you are ready! :)

Comment (matth79) [reply acoustic/flatten_bird_csv/src/flatten_bird_csv.py]: Yes, that's kind of what I was getting at with the ReadableFile thing. It worked and is in the later commits.

Comment (sdenton4) [reply acoustic/flatten_bird_csv/src/flatten_bird_csv.py]: Gotcha; sorry for the noise.

**Comment by matth79** [comment]
This is just meant to have BigQuery "on hand" in case it may prove useful, not as a replacement for the output of readML.R.

The permuted class index problem is fixed. Please check the updated dashboard: https://datastudio.google.com/c/u/0/reporting/8495dcff-77c5-4a42-812e-7ae5ed79494c/page/sFToC

(...and the BigQuery tables in the comb datasource, if you know the project id.)

Dataflow jobs have been often but not always getting stuck on startup, but we could debug that later. Local runs with parallel workers are working fine. With min_logit=-1, it can run in under 1 hour on my laptop.

**Comment by ddkapan** [comment]
Thanks Matt! This is great to have the ability to query the very large dataset and push it to your  dynamic map.  See you tomorrow! Durrell
